### DISCUSSION TITLE
关于过滤器设计问题

### DISCUSSION THREAD

**Discussion by vearne** [apache/rocketmq#6770] [General]
当前tag类型的过滤表达式是
1）在isMatchedByConsumeQueue() 使用Set<Integer> codeSet做粗过滤
2）在isMatchedByCommitLog()是直接返回true(等于什么都没干)
3）然后在client再按照Set<String> tagsSet再做细过滤

我的疑问是为什么不干脆在isMatchedByCommitLog()使用Set<String> tagsSet进行细过滤，
杜绝无效的数据在broker和client端传递的可能

[ExpressionMessageFilter.java](https://github.com/apache/rocketmq/blob/develop/broker/src/main/java/org/apache/rocketmq/broker/filter/ExpressionMessageFilter.java#L117)

**Comment by drpmma** [comment]
这个设计的出发点可能不希望Broker CPU过载过高，这是一个吞吐和CPU的权衡，而且codeSet应该能应付大多数情况了。
不过确实可以考虑直接在Broker做完过滤，是可以进一步讨论下。

The underlying intention of this design may be to avoid overloading the Broker CPU, which is a trade-off between throughput and CPU usage. Additionally, the codeSet should be able to handle most situations.
However, it is indeed worth considering directly performing filtering on the Broker, and this is a topic that can be further discussed.

**Reply by vearne** [reply]
[PullAPIWrapper.processPullResult()](https://github.com/apache/rocketmq/blob/rocketmq-all-4.8.0/client/src/main/java/org/apache/rocketmq/client/impl/consumer/PullAPIWrapper.java#L70)
```
subscriptionData.getTagsSet().contains(msg.getTags()
```
处理逻辑非常简单，几乎没有什么开销。
如果Broker都过滤干净，client的代码还能进一步精简
